Fix #5176 by premultiplying the alpha channel to the color channels

hirschenberger opened the pull request:
This is GIF specific. It's also done when the image is PNG but PNG is
handled separately with the PNG crate, whereas GIFs are handled by the
stb-image crate and the distinction between alpha and non-alpha-supporting
images was missing.

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @larsbergstrom (or someone else) soon.

hirschenberger committed:
Fix #5176 by premultiplying the alpha channel to the color channels

This is GIF specific. It's also done when the image is PNG but PNG is
handled separately with the PNG crate, whereas GIFs are handled by the
stb-image crate and the distinction between alpha and non-alpha-supporting
images was missing.

larsbergstrom commented:
Thanks for providing this fix!

Do you have some ideas for how we could add a small reftest, too? We have instructions at: https://github.com/servo/servo/wiki/Testing#reftests , but the general idea is that you would commit two versions of the page, perhaps one with a small gif that is empty (just alpha) and a blank page and that they are the same, which would ensure that we don't regress this feature.

CC: @metajack to make sure that reftest idea isn't stupid

hirschenberger committed:
Add reftest for issue #5176

hirschenberger commented:
I added a test similar to the PNG alpha channel reftest

larsbergstrom commented:
@hirschenberger Awesome - looks great!
